File: CyxbsMobile2019_iOS/Rising Schedule/Layout Support/ScheduleCollectionViewLayout.m
//
//  ScheduleCollectionViewLayout.m
//  CyxbsMobile2019_iOS
//
//  Created by SSR on 2022/9/4.
//  Copyright © 2022 Redrock. All rights reserved.
//

#import "ScheduleCollectionViewLayout.h"

#pragma mark - ScheduleCollectionViewLayout ()

@interface ScheduleCollectionViewLayout ()

/// 获取section的值
@property (nonatomic) NSInteger sections;

/// 正常视图布局
@property (nonatomic, strong) NSMutableDictionary <NSIndexPath *, ScheduleCollectionViewLayoutAttributes *> *itemAttributes;

/// 补充视图布局
@property (nonatomic, strong) NSMutableDictionary <NSString *, NSMutableDictionary <NSIndexPath *, UICollectionViewLayoutAttributes *> *> *supplementaryAttributes;

@end

#pragma mark - ScheduleCollectionViewLayout

@implementation ScheduleCollectionViewLayout

- (instancetype)init {
    self = [super init];
    if (self) {
        _itemAttributes = NSMutableDictionary.dictionary;
        _supplementaryAttributes = @{
            UICollectionElementKindSectionHeader : NSMutableDictionary.dictionary,
            UICollectionElementKindSectionLeading : NSMutableDictionary.dictionary,
            UICollectionElementKindSectionPlaceholder : NSMutableDictionary.dictionary
        }.mutableCopy;
    }
    return self;
}

#pragma mark - Ask LayoutAttributes

- (NSArray<__kindof UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect {
    NSMutableArray *result = NSMutableArray.array;
    for (NSInteger section = 0; section < _sections; section++) {
        
        // SupplementaryView attributes
        for (NSString *elementKind in _supplementaryAttributes.allKeys) {
            id <ScheduleCollectionViewDataSource> dataSource = (id <ScheduleCollectionViewDataSource>)self.collectionView.dataSource;
            NSInteger supplementaryCount = ![dataSource respondsToSelector:@selector(collectionView:numberOfSupplementaryOfKind:inSection:)] ? 0 :
                [dataSource collectionView:self.collectionView numberOfSupplementaryOfKind:elementKind inSection:section];
            for (NSInteger item = 0; item < supplementaryCount; item++) {
                
                NSIndexPath *indexPath = [NSIndexPath indexPathForItem:item inSection:section];
                UICollectionViewLayoutAttributes *attributes = [self layoutAttributesForSupplementaryViewOfKind:elementKind atIndexPath:indexPath];
                if (CGRectIntersectsRect(rect, attributes.frame)) {
                    [result addObject:attributes];
                }
            }
        }
        
        // Cell attributes
        NSUInteger itemCount = [self.collectionView.dataSource collectionView:self.collectionView numberOfItemsInSection:section];
        for (NSInteger item = 0; item < itemCount; item++) {
            
            NSIndexPath *indexPath = [NSIndexPath indexPathForItem:item inSection:section];
            UICollectionViewLayoutAttributes *attributes = [self layoutAttributesForItemAtIndexPath:indexPath];
            if (CGRectIntersectsRect(rect, attributes.frame)) {
                [result addObject:attributes];
            }
        }
    }
    
    return result;
}

// --------------- Item ---------------

- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath {
    NSParameterAssert(indexPath);
    
    ScheduleCollectionViewLayoutAttributes *attributes = _itemAttributes[indexPath];
    if (attributes) {
        NSAssert(attributes.pointIndexPath.week, @"week为0");
        return attributes;
    }

    attributes = [self.class.layoutAttributesClass layoutAttributesForCellWithIndexPath:indexPath];
    _itemAttributes[indexPath] = attributes;
    
    if (self.dataSource) {
        NSIndexPath *locationIndexPath = [self.dataSource collectionView:self.collectionView layout:self locationAtIndexPath:indexPath];
        NSInteger lenth = [self.dataSource collectionView:self.collectionView layout:self lenthForLocationIndexPath:locationIndexPath];
        attributes.pointIndexPath = locationIndexPath;
        attributes.lenth = lenth;
        
        [self _transformItemWithAttributes:attributes];
    }
    
    return attributes;
}

- (void)_transformItemWithAttributes:(ScheduleCollectionViewLayoutAttributes *)attributes {
    
    CGFloat x = attributes.pointIndexPath.section * self.collectionView.width + self.widthForLeadingSupplementaryView + (attributes.pointIndexPath.week - 1) * (_itemSize.width + self.columnSpacing);
    CGFloat y = self.heightForHeaderSupplementaryView + (attributes.pointIndexPath.location - 1) * (_itemSize.height + self.lineSpacing) + self.lineSpacing;
    CGFloat height = attributes.lenth * _itemSize.height + (attributes.lenth - 1) * self.columnSpacing;

    CGRect frame = CGRectMake(x, y, _itemSize.width, height);
    
    attributes.frame = frame;
}

// --------------- Supplementary ---------------

- (UICollectionViewLayoutAttributes *)layoutAttributesForSupplementaryViewOfKind:(NSString *)elementKind atIndexPath:(NSIndexPath *)indexPath {
    NSParameterAssert(indexPath);
    
    UICollectionViewLayoutAttributes *attributes = _supplementaryAttributes[elementKind][indexPath];

    if (!attributes) {
        attributes = [UICollectionViewLayoutAttributes layoutAttributesForSupplementaryViewOfKind:elementKind withIndexPath:indexPath];
        _supplementaryAttributes[elementKind][indexPath] = attributes;
    }
    
    [self _transformSupplementaryAttributes:attributes];
    
    return attributes;
}

- (void)_transformSupplementaryAttributes:(UICollectionViewLayoutAttributes *)attributes {
    NSIndexPath *indexPath = attributes.indexPath;
    NSString *elementKind = attributes.representedElementKind;
    
    // Header Element
    if ([elementKind isEqualToString:UICollectionElementKindSectionHeader]) {
        if (indexPath.item == 0) {
            CGFloat x = indexPath.section * self.collectionView.width;
            CGFloat y = self.collectionView.contentOffset.y;
            CGRect frame = CGRectMake(x, y, self.widthForLeadingSupplementaryView, self.heightForHeaderSupplementaryView);
            
            attributes.frame = frame;
            attributes.zIndex = 100;
            
            return;
        }
        CGFloat x = indexPath.section * self.collectionView.width + self.widthForLeadingSupplementaryView + (indexPath.item - 1) * (self.columnSpacing + _itemSize.width);
        CGFloat y = self.collectionView.contentOffset.y;
        
        CGRect frame = CGRectMake(x, y, _itemSize.width, self.heightForHeaderSupplementaryView);
        
        attributes.frame = frame;
        attributes.zIndex = 100;
        
        return;
    }
    
    // Leading Element
    if ([elementKind isEqualToString:UICollectionElementKindSectionLeading]) {
        CGFloat x = indexPath.section * self.collectionView.width;
        CGFloat y = self.heightForHeaderSupplementaryView + indexPath.item * (self.lineSpacing + _itemSize.height);
        
        CGRect frame = CGRectMake(x, y, self.widthForLeadingSupplementaryView, _itemSize.height);
        
        attributes.frame = frame;
        
        return;
    }
    
    // Placeholder Element
    if ([elementKind isEqualToString:UICollectionElementKindSectionPlaceholder]) {
        CGFloat x = indexPath.section * self.collectionView.width + self.widthForLeadingSupplementaryView;
        CGFloat y = self.collectionView.contentOffset.y;// + self.heightForHeaderSupplementaryView;
        CGFloat width = self.collectionView.width - self.widthForLeadingSupplementaryView;
        CGFloat height = self.collectionView.height - self.heightForHeaderSupplementaryView;
        
        CGRect frame = CGRectMake(x, y, width, height);
        
        attributes.frame = frame;
        
        return ;
    }
}

#pragma mark - Method

- (NSIndexPath *)indexPathAtPoint:(CGPoint)point {
    NSInteger section = point.x / self.collectionView.width;
    NSInteger week = (point.x - section * self.collectionView.width - self.widthForLeadingSupplementaryView) / (self.itemSize.width + self.columnSpacing) + 1;
    NSInteger location = (point.y + self.collectionView.contentOffset.y) / (self.itemSize.height + self.lineSpacing);
    
    return [NSIndexPath indexPathForLocation:location inWeek:week inSection:section];
}

#pragma mark - Others

- (void)prepareLayout {
    [self _calculateLayoutIfNeeded];
}

- (BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds {
    return YES;
}

- (CGSize)collectionViewContentSize {
    
    NSInteger itemCount = 13;
    
    CGSize contentSize = CGSizeMake(_sections * self.collectionView.bounds.size.width, self.heightForHeaderSupplementaryView + itemCount * (_itemSize.height + self.lineSpacing));
    
    return contentSize;
}

//- (CGPoint)targetContentOffsetForProposedContentOffset:(CGPoint)proposedContentOffset withScrollingVelocity:(CGPoint)velocity {
//    CGPoint point = [super targetContentOffsetForProposedContentOffset:proposedContentOffset withScrollingVelocity:velocity];
//    NSLog(@"from point:%@ with velocity:%@ to:%@", NSStringFromCGPoint(proposedContentOffset), NSStringFromCGPoint(velocity), NSStringFromCGPoint(point));
//    return point;
//}

#pragma mark - (UISubclassingHooks)

+ (Class)layoutAttributesClass {
    return ScheduleCollectionViewLayoutAttributes.class;
}

+ (Class)invalidationContextClass {
    return ScheduleCollectionViewLayoutInvalidationContext.class;
}

- (UICollectionViewLayoutInvalidationContext *)invalidationContextForBoundsChange:(CGRect)newBounds {
    
    ScheduleCollectionViewLayoutInvalidationContext *context =
    (ScheduleCollectionViewLayoutInvalidationContext *)[super invalidationContextForBoundsChange:newBounds];
    
    context.invalidateHeaderSupplementaryAttributes = YES;
    
    return context;
}

- (void)invalidateLayoutWithContext:(ScheduleCollectionViewLayoutInvalidationContext *)context {
    
    // invalidate Header Supplementary
    if (context.invalidateHeaderSupplementaryAttributes) {
        
        [_supplementaryAttributes[UICollectionElementKindSectionHeader] enumerateKeysAndObjectsUsingBlock:^(NSIndexPath * _Nonnull indexPath, UICollectionViewLayoutAttributes * _Nonnull attributes, BOOL * __unused stop) {
            
            CGFloat x = indexPath.section * self.collectionView.width;
            CGFloat y = self.collectionView.contentOffset.y;
            
            CGRect frame = CGRectMake(x, y, attributes.size.width, attributes.size.height);
            
            attributes.frame = frame;
        }];
        
        [_supplementaryAttributes[UICollectionElementKindSectionPlaceholder] enumerateKeysAndObjectsUsingBlock:^(NSIndexPath * _Nonnull indexPath, UICollectionViewLayoutAttributes * _Nonnull attributes, BOOL * __unused stop) {
            
            CGFloat x = indexPath.section * self.collectionView.width + self.widthForLeadingSupplementaryView;
            CGFloat y = self.collectionView.contentOffset.y + self.heightForHeaderSupplementaryView;
            
            CGRect frame = CGRectMake(x, y, attributes.size.width, attributes.size.height);
            
            attributes.frame = frame;
        }];
    }
    
    // invalidate Leading Supplementary
    if (context.invalidateLeadingSupplementaryAttributes) {
        
        [_supplementaryAttributes[UICollectionElementKindSectionLeading] enumerateKeysAndObjectsUsingBlock:^(NSIndexPath * _Nonnull indexPath, UICollectionViewLayoutAttributes * _Nonnull attributes, BOOL * __unused stop) {
            
            CGFloat x = indexPath.section * self.collectionView.width;
            CGFloat height = (_itemSize.height + self.lineSpacing) * 12;
            
            CGRect frame = CGRectMake(x, self.heightForHeaderSupplementaryView, self.widthForLeadingSupplementaryView, height);
            
            attributes.frame = frame;
        }];
    }
    
    // invalidate All Attributes
    if (context.invalidateDataSourceCounts) {
        
        [_itemAttributes removeAllObjects];
        
        [_supplementaryAttributes enumerateKeysAndObjectsUsingBlock:^(NSString * __unused key, NSMutableDictionary<NSIndexPath *,UICollectionViewLayoutAttributes *> * _Nonnull obj, BOOL * __unused stop) {
            [obj removeAllObjects];
        }];
    }
    
    [super invalidateLayoutWithContext:context];
}

#pragma mark - Private API

- (void)_calculateLayoutIfNeeded {
    
    NSInteger sections = [self.collectionView.dataSource performSelector:@selector(numberOfSectionsInCollectionView:) withObject:self.collectionView]
    ? [self.collectionView.dataSource numberOfSectionsInCollectionView:self.collectionView]
    : [self.collectionView numberOfSections];
    
    _sections = sections;
    
    CGFloat width = (self.collectionView.bounds.size.width - self.widthForLeadingSupplementaryView) / 7 - self.columnSpacing;
    
    _itemSize = CGSizeMake(width, width / 46 * 50);
}

@end

File: CyxbsMobile2019_iOS/Discover/Discover_WeDate/CourseSchedule/ScheduleCollectionViewLayout.swift
//
//  ScheduleCollectionViewLayout.swift
//  CyxbsMobile2019_iOS
//
//  Created by coin on 2023/9/17.
//  Copyright © 2023 Redrock. All rights reserved.
//

import UIKit

// MARK: ScheduleCollectionViewLayoutDataSource

@objc public
protocol ScheduleCollectionViewLayoutDataSource {
    
    func collectionView(_ collectionView: UICollectionView, layout: ScheduleCollectionViewLayout, columnOfItemAtIndexPath indexPath: IndexPath) -> Int
    
    func collectionView(_ collectionView: UICollectionView, layout: ScheduleCollectionViewLayout, lineOfItemAtIndexPath indexPath: IndexPath) -> Int
    
    func collectionView(_ collectionView: UICollectionView, layout: ScheduleCollectionViewLayout, lenthOfItemAtIndexPath indexPath: IndexPath) -> Int
    
    func collectionView(_ collectionView: UICollectionView, layout: ScheduleCollectionViewLayout, numberOfSupplementaryOfKind kind: String, inSection section: Int) -> Int
    
    func collectionView(_ collectionView: UICollectionView, layout: ScheduleCollectionViewLayout, persentOfPointAtIndexPath indexPath: IndexPath) -> CGFloat
}

// MARK: UICollectionView.ElementKindSection

extension UICollectionView {

    public enum ElementKindSection: String {

        case header = "Redrock.UICollectionView.ElementKindSection.header"

        case leading = "Redrock.UICollectionView.ElementKindSection.leading"

//        case placeHolder = "Redrock.UICollectionView.ElementKindSection.placeHolder"

//        case pointHolder = "Redrock.UICollectionView.ElementKindSection.pointHolder"
    }

    public func register(_ viewClass: AnyClass?, forElementKindSection elementKind: ElementKindSection, withReuseIdentifier identifier: String) {
        register(viewClass, forSupplementaryViewOfKind: elementKind.rawValue, withReuseIdentifier: identifier)
        if let layout = collectionViewLayout as? ScheduleCollectionViewLayout {
            layout.register(viewClass, forSupplementaryViewOfKind: elementKind.rawValue)
        }
    }
}

// MARK: ScheduleCollectionViewLayout

open class ScheduleCollectionViewLayout: UICollectionViewLayout {
    
    // MARK: Property
    
    open weak var dataSource: ScheduleCollectionViewLayoutDataSource?
    
    open var lineSpacing: CGFloat = 2
    
    open var columnSpacing: CGFloat = 2
    
    open var widthForLeadingSupplementaryView: CGFloat = 30
    
    private(set) var heightForHeaderSupplementaryView: CGFloat = 0

    open var aspectRatio: CGFloat = 46.0 / 50.0
    
    open var heightForHeader: CGFloat = 50
    
    open var pageCalculation: Int = 0
    
    private(set) open var itemSize: CGSize = .zero
    
    open var pageShows: Int = 1 {
        didSet {
            if pageShows < 1 {
                pageShows = 1
            }
        }
    }
    
    // MARK: Private
    
    private var itemAttributes: [IndexPath: UICollectionViewLayoutAttributes] = [:]
    
    private var supplementaryAttributes: [String: [IndexPath: UICollectionViewLayoutAttributes]] = [:]
    
    // MARK: Method - middle
    
    open func register(_ viewClass: AnyClass?, forSupplementaryViewOfKind elementKind: String) {
        if supplementaryAttributes[elementKind] == nil {
            supplementaryAttributes[elementKind] = [:]
        }
    }
    
    // MARK: Method - layoutAttributes
    
    open override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        let numberOfSections = collectionView?.dataSource?.numberOfSections?(in: collectionView!) ?? 0
        if numberOfSections <= 0 { return nil }
        
        var result = [UICollectionViewLayoutAttributes]()
        for section in 0 ..< numberOfSections {
            for elementKind in supplementaryAttributes.keys {
                let numberOfSupplementaryOfKind = dataSource?.collectionView(collectionView!, layout: self, numberOfSupplementaryOfKind: elementKind, inSection: section) ?? 0
                if numberOfSupplementaryOfKind <= 0 { continue }
                for item in 0..<numberOfSupplementaryOfKind {
                    let indexPath = IndexPath(item: item, section: section)
                    if let attributes = layoutAttributesForSupplementaryView(ofKind: elementKind, at: indexPath) {
                        if rect.intersects(attributes.frame) {
                            result.append(attributes)
                        }
                    }
                }
            }
            let itemCount = collectionView?.dataSource?.collectionView(collectionView!, numberOfItemsInSection: section) ?? 0
            if itemCount <= 0 { continue }
            for item in 0 ..< itemCount {
                let indexPath = IndexPath(item: item, section: section)
                if let attributes = layoutAttributesForItem(at: indexPath) {
                    if rect.intersects(attributes.frame) {
                        result.append(attributes)
                    }
                }
            }
        }
        return result
    }
    
    open override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        guard let dataSource = dataSource else { return nil }
        let columnOfItem = dataSource.collectionView(collectionView!, layout: self, columnOfItemAtIndexPath: indexPath)
        let lineOfItem = dataSource.collectionView(collectionView!, layout: self, lineOfItemAtIndexPath: indexPath)
        let lenthOfItem = dataSource.collectionView(collectionView!, layout: self, lenthOfItemAtIndexPath: indexPath)
        
        let attribute = itemAttributes[indexPath] ?? UICollectionViewLayoutAttributes(forCellWith: indexPath)
        attribute.frame = CGRect(
            x: CGFloat(indexPath.section) * collectionView!.bounds.width + widthForLeadingSupplementaryView + CGFloat(columnOfItem - 1) * (itemSize.width + columnSpacing),
            y: heightForHeaderSupplementaryView + CGFloat(lineOfItem - 1) * (itemSize.height + lineSpacing) + lineSpacing,
            width: itemSize.width,
            height: CGFloat(lenthOfItem) * itemSize.height + CGFloat(lenthOfItem - 1) * columnSpacing)
        itemAttributes[indexPath] = attribute
        
        return attribute
    }
    
    open override func layoutAttributesForSupplementaryView(ofKind elementKind: String, at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        guard let elementKindSection = UICollectionView.ElementKindSection(rawValue: elementKind) else { return nil }
        guard let elementKindAttributes = supplementaryAttributes[elementKind] else { return nil }
        let attributes = elementKindAttributes[indexPath] ?? UICollectionViewLayoutAttributes(forSupplementaryViewOfKind: elementKind, with: indexPath)
        setupSupplementaryAttributes(attributes, elementKindSection: elementKindSection)
        supplementaryAttributes[elementKind]?[indexPath] = attributes
        return attributes
    }
    
    func setupSupplementaryAttributes(_ attributes :UICollectionViewLayoutAttributes, elementKindSection: UICollectionView.ElementKindSection) {
        
        let indexPath = attributes.indexPath
        let pageWidth = collectionView!.bounds.width / CGFloat(pageShows)
        
        switch elementKindSection {
        case .header:
            attributes.zIndex = 10
            
            if indexPath.item == 0 {
                attributes.frame = CGRect(
                    x: CGFloat(indexPath.section) * pageWidth,
                    y: collectionView!.contentOffset.y,
                    width: widthForLeadingSupplementaryView,
                    height: heightForHeaderSupplementaryView)
            } else {
                attributes.frame = CGRect(
                    x: CGFloat(indexPath.section) * collectionView!.frame.width + widthForLeadingSupplementaryView + CGFloat(indexPath.item - 1) * (columnSpacing + itemSize.width),
                    y: collectionView!.contentOffset.y,
                    width: itemSize.width,
                    height: heightForHeaderSupplementaryView)
            }
            
        case .leading:
            attributes.frame = CGRect(
                x: CGFloat(indexPath.section) * collectionView!.frame.width,
                y: heightForHeaderSupplementaryView + CGFloat(indexPath.item) * (lineSpacing + itemSize.height),
                width: widthForLeadingSupplementaryView,
                height: itemSize.height)
            
//        case .pointHolder:
////            let persent = dataSource?.collectionView(ry_collectionView, layout: self, persentOfPointAtIndexPath: indexPath) ?? 0
//            print("todo")
        }
    }
    
    // MARK: others
    
    open override func prepare() {
        calculateLayoutIfNeeded()
    }
    
    open override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
        true
    }
    
    open override func invalidateLayout(with context: UICollectionViewLayoutInvalidationContext) {
        if context.invalidateDataSourceCounts {
            itemAttributes.removeAll(keepingCapacity: true)
            for key in supplementaryAttributes {
                supplementaryAttributes[key.key]?.removeAll(keepingCapacity: true)
            }
        }
        supplementaryAttributes[UICollectionView.ElementKindSection.header.rawValue]?.forEach { entry in
            self.setupSupplementaryAttributes(entry.value, elementKindSection: .header)
        }
    }
    
    open override var collectionViewContentSize: CGSize {
        let itemCount = 13
        let sections = collectionView?.dataSource?.numberOfSections?(in: collectionView!) ?? 0
        return CGSize(
            width: CGFloat(sections) * collectionView!.bounds.width,
            height: heightForHeaderSupplementaryView + CGFloat(itemCount) * (itemSize.height + lineSpacing))
    }
    
    open override func targetContentOffset(forProposedContentOffset proposedContentOffset: CGPoint, withScrollingVelocity velocity: CGPoint) -> CGPoint {
        
        let toTime = Int(collectionView!.contentOffset.y / (itemSize.height + lineSpacing) + 0.5)
        let toY = (itemSize.height + lineSpacing) * CGFloat(toTime)
        
        var index = Int(proposedContentOffset.x / collectionView!.bounds.width + 0.5)
        let remainder = proposedContentOffset.x - CGFloat(index) * collectionView!.bounds.width
        
        if velocity.x > 0.6 || (velocity.x > 0.3 && remainder > collectionView!.bounds.width / 3) {
            index += 1
        }
        if velocity.x < -0.6 || (velocity.x < -0.3 && remainder < collectionView!.bounds.width / 3) {
            index -= 1
        }
        index = max(index, pageCalculation - 1)
        index = min(index, pageCalculation + 1)
        
        let toX = collectionView!.bounds.width * CGFloat(index)
        
        return CGPoint(x: toX, y: toY)
    }
    
    // MARK: private
    
    private func calculateLayoutIfNeeded() {
        let width = (collectionView!.bounds.width - widthForLeadingSupplementaryView) / 7 - columnSpacing
        itemSize = CGSize(width: width, height: width / aspectRatio)
        heightForHeaderSupplementaryView = width / aspectRatio
    }
}
